lib/providers/ipmeta_provider_maxmind_v2.c

```diff
@@ -0,0 +1,914 @@
+  state->current_line++;
+  /* reset the current column */
+  state->current_column = 0;
+  /* reset the temp record */
+  memset(&(state->tmp_record), 0, sizeof(ipmeta_record_t));
+  return;
+}
+
+/** Read a locations file */
+static int read_locations(ipmeta_provider_t *provider, io_t *file)
+{
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+
+  char buffer[BUFFER_LEN];
+  int read = 0;
+
+  /* reset the state variables before we start */
+  state->current_column = 0;
+  state->current_line = 0;
+  memset(&(state->tmp_record), 0, sizeof(ipmeta_record_t));
+
+  /* options for the csv parser */
+  int options = CSV_STRICT | CSV_REPALL_NL | CSV_STRICT_FINI | CSV_APPEND_NULL |
+                CSV_EMPTY_IS_NULL;
+
+  csv_init(&(state->parser), options);
+
+  while ((read = wandio_read(file, &buffer, BUFFER_LEN)) > 0) {
+    if (csv_parse(&(state->parser), buffer, read,
+                  parse_maxmind_v2_location_cell, parse_maxmind_v2_location_row,
+                  provider) != read) {
+      ipmeta_log(__func__, "Error parsing %s Location file", provider->name);
+      ipmeta_log(__func__, "CSV Error: %s",
+                 csv_strerror(csv_error(&(state->parser))));
+      return -1;
+    }
+  }
+
+  if (csv_fini(&(state->parser), parse_maxmind_v2_location_cell,
+               parse_maxmind_v2_location_row, provider) != 0) {
+    ipmeta_log(__func__, "Error parsing %s Location file", provider->name);
+    ipmeta_log(__func__, "CSV Error: %s",
+               csv_strerror(csv_error(&(state->parser))));
+    return -1;
+  }
+
+  csv_free(&(state->parser));
+
+  return 0;
+}
+
+/** Parse a blocks cell */
+static void parse_blocks_cell(void *s, size_t i, void *data)
+{
+  ipmeta_provider_t *provider = (ipmeta_provider_t *)data;
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+  ipmeta_record_t *tmp = &(state->tmp_record);
+  char *tok = (char *)s;
+  char *end;
+  char *mask_str;
+
+  /* skip the first lines */
+  if (state->current_line < HEADER_ROW_CNT) {
+    return;
+  }
+
+  switch (state->current_column) {
+  case BLOCKS_COL_NETWORK:
+    /* extract the mask from the prefix */
+    if ((mask_str = strchr(tok, '/')) != NULL) {
+      *mask_str = '\0';
+      mask_str++;
+      state->block_network.masklen = atoi(mask_str);
+    } else {
+      state->block_network.masklen = 32;
+    }
+
+    state->block_network.addr = inet_addr(tok);
+    if  (state->block_network.addr == INADDR_NONE){
+        ipmeta_log(__func__, "Invalid Start IP Value (%s)", tok);
+        state->parser.status = CSV_EUSER;
+    }
+    break;
+
+  case BLOCKS_COL_GEONAMEID:
+  /* Geoname ID*/
+    if (tok != NULL){
+      /* id */
+    tmp->id = strtol(tok, &end, 10);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid ID Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+    }
+    }
+    break;
+
+  case BLOCKS_COL_CCGEONAMEID:
+    /* Registered country geoname ID */
```

Comment:
please add a comment explaining why (and in what cases) the ID field is populated from the CCGEONAMEID column rather than GEONAMEID